### ISSUE TITLE
fresnelc/fresnels.as_real_imag with extended_real

### ISSUE THREAD

**Issue by oscargus** [sympy/sympy#16999]
```
from sympy import Symbol, fresnelc, fresnels
x = Symbol('x', extended_real=True)
fresnelc(x).as_real_imag()
```
gives 
`(fresnelc(fresnelc(x)), nan)`
but should be
`(fresnelc(x), 0)`

Similar for `fresnels`.

**Comment by oscarbenjamin** [comment]
It looks like `_as_real_imag` does the right thing but then there is an `as_real_imag` method that turns it into something else:
https://github.com/sympy/sympy/blob/5f2b57688b7a51cd8455055b35928320076a589f/sympy/functions/special/error_functions.py#L2062-L2074
Not sure why that method is needed...

**Comment by oscargus** [comment]
Probably the extended_real test should be executed before applying the equation? Or check if the imaginary part is zero before applying it?

The links goes to explicit equations for it which makes sense I guess, but they do not seem to have been implemented correctly.

(I guess it also makes sense to add these special equations in `_as_real_imag` and not overload the generic method.

**Comment by oscargus** [comment]
Or, I realize, do 

`x, y = self.args[0]._as_real_imag(deep=deep, **hints)`

(Was too focused on the testing to consider the actual bug...)

**Comment by jksuom** [comment]
I wonder why `x*sqrt(sq)` is used instead of `y`. It would be much easier to implement the imaginary part that way. (`x/(2*y) * sqrt(sq)` is just `sign(y)/2`.)

**Comment by oscargus** [comment]
As I understand it the sqrt part is `sqrt(-y**2/x**2) = I*abs(y)/abs(x)` so the real part has imaginary parts which (assumingly) cancel. But not sure. I seems to be correct based on the equations in the link (at least after a quick glance) but not intuitive at all.

**Comment by oscargus** [comment]
Now that I really read what you wrote...

Wouldn't `x*sqrt(sq) = I*abs(y)*sign(x)`?

**Comment by jksuom** [comment]
Yes, I carelessly ignored the `I` component. The formula is very general. If an analytic function `f` is real on the real axis, then it commutes with conjugation: `conjugate(f(z)) == f(conjugate(z))`. It follows that `re(f(x + I*y)` is `(f(x + I*y) + f(x - I*y))/2`  (regardless of `sign(y)`) and similarly for `im(f(z))`.

**Comment by jksuom** [comment]
`im(f(z) = (f(x + I*y) - f(x - I*y))/2/I`

**Comment by oscargus** [comment]
Thanks! I'll add this to #16996. (It seems like a rather general property that can be applied to a lot of functions, maybe one should make a class `RealOnRealFunction` or something and subclass this...)

**Comment by jksuom** [comment]
I think that even `RealFunction` would not be misleading.

**Comment by oscargus** [comment]
I made a temporary solution in #16996 using the same method for all error functions, `real_to_real_as_real_imag`, but in the long run, something like `RealFunction` should be implemented with the methods `as_real_imag` and `_eval_conjugate` (maybe something to evaluate `re` and `im`, although I cannot really find how that is done...).
